packages/taquito/src/contract/rpc-contract-provider.ts

```diff
@@ -124,61 +124,46 @@ export class RpcContractProvider extends OperationEmitter
    * @param keys Array of keys to query (will be encoded properly according to the schema)
    * @param schema Big Map schema (can be determined using your contract type)
    * @param block optional block level to fetch the values from
-   * @returns An object containing the keys queried in the big map and their value in a well-formatted JSON object format
+   * @returns A MichelsonMap containing the keys queried in the big map and their value in a well-formatted JSON object format
    *
    */
-  async getBigMapKeysByID<T>(id: string, keys: string[], schema: Schema, block?: number): Promise<{ [key: string]: T | undefined }> {
-    const bigMapValues = Object({[keys[0]]: undefined});
-    if (keys.length === 1) { // No need to get the block level if only one key
-      let val: T | undefined;
-      try {
-        val = await this.getBigMapKeyByID<T>(id, keys[0], schema, block);
-      } catch(ex) {
-        if (ex instanceof HttpResponseError && ex.status === STATUS_CODE.NOT_FOUND) {
-          val = undefined;
-        } else {
-          throw ex;
-        }
-      }
-      Object.assign(bigMapValues, {[keys[0]]: val});
+  async getBigMapKeysByID<T>(id: string, keys: Array<BigMapKeyType>, schema: Schema, block?: number): Promise<MichelsonMap<MichelsonMapKey, T | undefined>> {
+    const level = await this.getBlockForRequest(keys, block)
+    const bigMapValues = new MichelsonMap<MichelsonMapKey, T | undefined>();
+
+    // Execute batch of promises in series
+    const batchSize = 5;
+    let position = 0;
+    let results: Array<(T | undefined)> = [];
+
+    while (position < keys.length) {
+      const keysBatch = keys.slice(position, position + batchSize);
+      const batch = keysBatch.map((keyToEncode) => this.getBigMapValueOrUndefined<T>(keyToEncode, id, schema, level))
```

Comment:
Do we have a test for the negative case? If the key is not found we return undefined. What happens when there is an unexpected exception? Should we test that these exceptions bubble up to the caller?

Reply:
I added a test case, thank you!